### PR TITLE
Memory pool related leaks

### PR DESCRIPTION
**Pull request by bretambrose**
* Fixes a number of leaks (some may be only theoretically possible) related to not releasing messages on failures to aws_channel_slot_send_message() or the corresponding packet encode function.
* Fixes several leaks related to requests being uncompleted or completed but not yet removed when the connection is shut down.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/client_channel_handler.c [hunk 1]
```diff
@@ -602,8 +626,42 @@ static void s_request_timeout_task(struct aws_channel_task *task, void *arg, enu
...
+         * If the task is cancelled, assume safe shutdown is in progress.
+         * There are two distinct situations where requests end up with a cancelled task:
+         *   (1) The request has been completed but disconnect was begun before the task had a chance to run
+         *       normally.  We need a way to ensure that aws_memory_pool_release is called for this request.
+         *
+         *       We can do that by
+         *          (1) setting request->cancelled to true.  Then the hash table's value removal function will
+         *              ensure aws_memory_pool_release is called when the table is cleared during final disconnect.
+         *          (2) If we're uncomfortable (semantically) with a request that is both "completed" and "cancelled"
+         *              then we could alternatively call aws_memory_pool_release here.
```
Comment (graebm) [inline source/client_channel_handler.c]: This comment will confuse future generations. Be more clear which of these things we're doing, or remove the alternative explanation of how we could do it.

File: source/client_channel_handler.c [hunk 2]
```diff
@@ -602,8 +626,42 @@ static void s_request_timeout_task(struct aws_channel_task *task, void *arg, enu
...
+            request->cancelled = true;
+            if (!request->completed) {
+                AWS_LOGF_DEBUG(
+                    AWS_LS_MQTT_CLIENT,
+                    "static: task id %p, completing with interrupt request for packet id "
+                    "%" PRIu16 ".",
+                    (void *)task,
+                    request->message_id);
+                if (request->on_complete) {
+                    request->on_complete(
```
Comment (graebm) [inline source/client_channel_handler.c]: should we be setting `request->completed = true`?
since we're firing the on_complete callback?

Comment (bretambrose) [reply source/client_channel_handler.c]: True, task shouldn't get rerun but its safer to be sure.

File: source/client_channel_handler.c (hunk 1 above)
Comment (bretambrose) [reply source/client_channel_handler.c]: Yeah was probably better left as a review comment, but wanted people to read it and say which they preferred (if any).

File: source/mqtt.c [hunk 3]
```diff
@@ -151,6 +151,9 @@ bool aws_mqtt_is_valid_topic_filter(const struct aws_byte_cursor *topic_filter)
             AWS_DEFINE_ERROR_INFO_MQTT(
                 AWS_ERROR_MQTT_NO_TOPICS_FOR_RESUBSCRIBE,
                 "Resubscribe not necessary, client is not subscribed to any topics."),
+            AWS_DEFINE_ERROR_INFO_MQTT(
+                AWS_ERROR_MQTT_REQUEST_INTERRUPTED,
+                "Mqtt operation interrupted by connection shutdown"),
```
[the remark this answers is not in the document]
Comment (bretambrose) [reply source/mqtt.c]: That seems reasonable.

**Pull request merged**